cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -1131,150 +1224,220 @@ aggregate_reader_metadata::select_row_groups(
-        num_rows_per_source[src_idx] += rows_this_rg;
-      }
-    }
-  } else {
-    // Reset and recompute input row group count to adjust for num_rows and skip_rows. Here, the
-    // output from predicate pushdown was empty. i.e., no row groups filtered.
-    total_row_groups = 0;
-
-    size_type count = 0;
-    for (size_t src_idx = 0; src_idx < per_file_metadata.size(); ++src_idx) {
-      auto const& fmd = per_file_metadata[src_idx];
-      for (size_t rg_idx = 0;
-           rg_idx < fmd.row_groups.size() and count < rows_to_skip + rows_to_read;
-           ++rg_idx) {
-        auto const& rg             = fmd.row_groups[rg_idx];
-        auto const chunk_start_row = count;
-        count += rg.num_rows;
-        if (count > rows_to_skip || count == 0) {
-          // Keep this row group, increase count
-          total_row_groups++;
-
-          // start row of this row group adjusted with rows_to_skip
-          num_rows_per_source[src_idx] += count;
-          num_rows_per_source[src_idx] -=
-            (chunk_start_row <= rows_to_skip) ? rows_to_skip : chunk_start_row;
-
-          // We need the unadjusted start index of this row group to correctly initialize
-          // ColumnChunkDesc for this row group in create_global_chunk_info() and calculate
-          // the row offset for the first pass in compute_input_passes().
-          selection.emplace_back(rg_idx, chunk_start_row, src_idx);
-
-          // If page-level indexes are present, then collect extra chunk and page info.
-          // The page indexes rely on absolute row numbers, not adjusted for skip_rows.
-          column_info_for_row_group(selection.back(), chunk_start_row);
-        }
-        // Adjust the number of rows for the last source file.
-        if (count >= rows_to_skip + rows_to_read) {
-          num_rows_per_source[src_idx] -= count - rows_to_skip - rows_to_read;
-        }
-      }
-    }
+  // Track the starting row index of the current row group
+  size_type current_row_index = 0;
+
+  // Track the number of rows read so far (may be different from the current row index if row bounds
+  // were applied)
+  size_type total_selected_rows = 0;
+
+  // For each data source
+  std::for_each(thrust::counting_iterator<size_t>(0),
+                thrust::counting_iterator(current_row_group_indices.size()),
+                [&](auto const& src_idx) {
+                  auto const& file_metadata = per_file_metadata[src_idx];
+
+                  // For each row group in this data source
+                  std::for_each(
+                    current_row_group_indices[src_idx].begin(),
+                    current_row_group_indices[src_idx].end(),
+                    [&](auto const& rg_idx) {
+                      // Validate the row group index
+                      CUDF_EXPECTS(
+                        rg_idx >= 0 and std::cmp_less(rg_idx, file_metadata.row_groups.size()),
+                        "Invalid rowgroup index",
```

Comment:
either this is wrong or half of the variable names are wrong :)
```suggestion
                        "Invalid row group index",
```